sdk/program/src/sysvar/slot_hashes.rs

```diff
@@ -62,13 +71,96 @@ impl Sysvar for SlotHashes {
     }
 }
 
+#[derive(Copy, Clone, Pod, Zeroable)]
+#[repr(C)]
+struct PodSlotHash {
+    slot: Slot,
+    hash: Hash,
+}
+
+/// Trait for querying the `SlotHashes` sysvar.
+pub trait SlotHashesSysvar {
+    /// Get a value from the sysvar entries by its key.
+    /// Returns `None` if the key is not found.
+    fn get(slot: &Slot) -> Result<Option<Hash>, ProgramError> {
+        let data_len = SlotHashes::size_of();
+        let mut data = vec![0u8; data_len];
```

Comment:
I think you'll need to do the alloc here instead of creating a stack variable, otherwise you'll blow out the stack on BPF when trying to access memory outside of the 4096

Reply:
I was under the impression that vectors are allocated on the heap by default.

https://doc.rust-lang.org/beta/alloc/vec/struct.Vec.html#guarantees

> If a Vec has allocated memory, then the memory it points to is on the heap (as defined by the allocator Rust is configured to use by default)

I'm assuming because we use the bump allocator this is different?

Reply:
Sorry, you're absolutely right. I honestly didn't know what would happen with a `vec!` that was instantiated with a huge value, so I did a little test:

```
use solana_program::{entrypoint::HEAP_LENGTH, msg};
#[no_mangle]
pub unsafe extern "C" fn entrypoint(_: *mut u8) -> u64 {
    const BIG_SIZE: usize = HEAP_LENGTH - 100;
    let mut big_vec = vec![0u8; BIG_SIZE];
    big_vec[BIG_SIZE - 1] = 1;
    msg!("{} size {}", big_vec[BIG_SIZE - 1], big_vec.len());
    0
}
solana_program::custom_panic_default!();
solana_program::custom_heap_default!();
```

And this works perfectly -- it didn't work with exactly `HEAP_LENGTH` probably due to allocations during `msg!`, so I lazily just put `HEAP_LENGTH - 100` which did work.

Reply:
It looks like `SlotHistory` is the only sysvar whose data could be larger than the heap, but it's actually omitted from `sol_get_sysvar` (intentionally), so the next largest sysvar that this API could handle is `SlotHashes` at `20_488`.

That means programs could allocate up to `12_280` of heap before calling either search method. Do you think this is acceptable, considering the vector allocated is dropped after the search concludes? Or should we "chunk" up the data load from `sol_get_sysvar` to allocate less heap? We could do something like `SloHashes::size_of() / 4` which is only `5_122` heap bytes allocated at a time.

Reply:
I haven't looked at how it's used, but at worst, it should be fine to make the syscall only once and carry around the data for the rest of the instruction if we need to use it more than once. If we end up blowing the heap, we're probably doing too many allocations elsewhere, and we can address that

Reply:
Okay, sounds reasonable!